drivers/ieee802154/ieee802154_fsci.c

```diff
@@ -0,0 +1,705 @@
+			addr_field->plain.addr.short_addr = data->srcAddr;
+		}
+
+		buf += IEEE802154_SHORT_ADDR_LENGTH;
+	} else { /* gAddrModeExtendedAddress_c */
+		if (fs->fc.pan_id_comp) {
+			memcpy(addr_field->comp.addr.ext_addr,
+			       (u8_t *)&data->srcAddr,
+			       IEEE802154_EXT_ADDR_LENGTH);
+		} else {
+			memcpy(addr_field->plain.addr.ext_addr,
+			       (u8_t *)&data->srcAddr,
+			       IEEE802154_EXT_ADDR_LENGTH);
+		}
+
+		buf += IEEE802154_EXT_ADDR_LENGTH;
+	}
+
+	msdu = fsci_context_data.rx_buf + sizeof(mcpsDataInd_t) -
+	       sizeof(u8_t *);
+	memcpy(buf, msdu, data->msduLength);
+	buf += data->msduLength;
+
+	net_buf_add(frag, (u8_t)(buf - frag->data));
+
+	if (ieee802154_radio_handle_ack(fsci->iface, pkt) == NET_OK) {
+		SYS_LOG_DBG("ACK packet handled");
+		goto out;
+	}
+
+	if (net_recv_data(fsci->iface, pkt) < 0) {
+		SYS_LOG_DBG("Packet dropped by NET stack");
+		goto out;
+	}
+
+	return;
+out:
+	if (pkt) {
+		net_pkt_unref(pkt);
+	}
+}
+
+static void fsci_rx_resp(void)
+{
+	memcpy(fsci_context_data.fsci_pkt.structured.payload,
+	       fsci_context_data.rx_buf,
+	       fsci_context_data.fsci_pkt.structured.header.len);
+
+	fsci_context_data.req_started = false;
+	k_sem_give(&fsci_context_data.req_sem);
+}
+
+static u8_t *fsci_rx(u8_t *buf, size_t *off)
+{
+	clientPacketStructured_t *pkt =
+		 &fsci_context_data.fsci_pkt.structured;
+
+	switch (fsci_context_data.rx_state) {
+	case FSCI_RX_STATE_SYNC:
+		if (*buf == 0x02) {
+			pkt->header.startMarker = *buf;
+			fsci_context_data.rx_state = FSCI_RX_STATE_OG;
+		}
+
+		break;
+	case FSCI_RX_STATE_OG:
+		pkt->header.opGroup = *buf;
+		fsci_context_data.rx_state = FSCI_RX_STATE_OC;
+		break;
+	case FSCI_RX_STATE_OC:
+		pkt->header.opCode = *buf;
+		fsci_context_data.rx_state = FSCI_RX_STATE_LEN;
+		break;
+	case FSCI_RX_STATE_LEN:
+		if (*buf > gFsciMaxPayloadLen_c) {
+			SYS_LOG_ERR("Payload buffer size not big enough! "
+				    " Needed size is %d", *buf);
+		}
+
+		pkt->header.len = *buf;
+		fsci_context_data.rx_off = 0;
+		fsci_context_data.rx_state = FSCI_RX_STATE_PAYLOAD;
+
+		break;
+	case FSCI_RX_STATE_PAYLOAD:
+		if (fsci_context_data.rx_off < gFsciMaxPayloadLen_c) {
+			fsci_context_data.rx_buf[
+				fsci_context_data.rx_off] = *buf;
+		}
+
+		fsci_context_data.rx_off++;
+
+		if (fsci_context_data.rx_off == pkt->header.len) {
+			fsci_context_data.rx_state = FSCI_RX_STATE_CRC;
+		}
+
+		break;
+	case FSCI_RX_STATE_CRC:
+		pkt->payload[pkt->header.len] = *buf;
+
+		if (pkt->header.opGroup == gFSCI_McpsNwkOpcodeGroup_c &&
+		    pkt->header.opCode == mFsciMcpsNwkDataInd_c) {
+			fsci_rx_data();
+		} else {
+			if (!fsci_context_data.req_started) {
+				SYS_LOG_ERR("Unexpected packet type: "
+					    "og=0x%02X oc=0x%02X",
+					    pkt->header.opGroup,
+					    pkt->header.opCode);
+			}
+
+			fsci_rx_resp();
+		}
+
+		fsci_context_data.rx_state = FSCI_RX_STATE_SYNC;
+		break;
+	}
+
+	*off = 0;
+	return buf;
+}
+
+static void fsci_fill_crc(clientPacket_t *pkt)
```

Comment:
the hardware is not able to fill this in by itself?
Even the cheapest 15.4 crappy chip usually does.